File: 2024/src/day02.py
#!/usr/bin/env python3

with open("inputs/02.txt", "r") as f:
    lines = f.readlines()

seqs = [list(map(int, line.split())) for line in lines]

# part a
valid_sols = 0
for seq in seqs:
    prev = seq[0]
    decreasing = seq[1] < seq[0]
    valid = True
    for el in seq[1:]:
        diff = el - prev
        if abs(diff) > 3 or diff == 0 or (diff < 0) != decreasing:
            valid = False
            break
        prev = el
    if valid:
        valid_sols += 1

print(valid_sols)


# part b
def check_valid(
    seq: list[int], i: int, prev_i: int, decreasing: bool, forgived
) -> bool:
    if i >= len(seq):
        return True

    diff = seq[i] - seq[prev_i]

    if abs(diff) > 3 or diff == 0 or (diff < 0) != decreasing:
        if not forgived:
            match prev_i:
                case 0:
                    return check_valid(
                        seq, i + 1, prev_i, seq[i + 1] < seq[prev_i], True
                    ) or check_valid(seq, i + 1, i, seq[i + 1] < seq[i], True)
                case 1:
                    return (
                        check_valid(seq, i + 1, prev_i, decreasing, True)
                        or check_valid(seq, i, i - 2, decreasing, True)
                        or check_valid(seq, i, prev_i, seq[i] < seq[prev_i], True)
                    )
                case _:
                    return check_valid(seq, i + 1, prev_i, decreasing, True) or (
                        check_valid(seq, i, i - 2, decreasing, True)
                    )
        else:
            return False
    else:
        return check_valid(seq, i + 1, i, decreasing, forgived)


print(sum(check_valid(seq, 1, 0, seq[1] < seq[0], False) for seq in seqs))

File: 2024/src/day02.rs
use std::fs;

pub fn solve() {
    let input = parse_input(fs::read_to_string("inputs/02.txt").unwrap());
    println!("Day 01:");
    println!("{}", solve_part_a(&input));
    println!("{}", solve_part_b(&input));
}

fn parse_input(input: String) -> Vec<Vec<i32>> {
    input
        .lines()
        .map(|line| line.split(" ").map(|num| num.parse().unwrap()).collect())
        .collect()
}

fn solve_part_a(input: &Vec<Vec<i32>>) -> usize {
    solve_day2(input, false)
}

fn solve_part_b(input: &Vec<Vec<i32>>) -> usize {
    solve_day2(input, true)
}
fn solve_day2(input: &Vec<Vec<i32>>, can_forgive: bool) -> usize {
    input
        .iter()
        .map(|seq| {
            if check_valid(seq, 1, 0, seq[1] < seq[0], can_forgive) {
                1
            } else {
                0
            }
        })
        .sum()
}

fn check_valid(
    seq: &Vec<i32>,
    i: usize,
    prev_i: usize,
    decreasing: bool,
    can_forgive: bool,
) -> bool {
    if i >= seq.len() {
        return true;
    }

    let diff = seq[i] - seq[prev_i];
    if diff.abs() > 3 || diff == 0 || (diff < 0) != decreasing {
        if can_forgive {
            match i {
                1 => {
                    return check_valid(seq, i + 1, i - 1, seq[i + 1] < seq[i - 1], false) // skip i
                        || check_valid(seq, i + 1, i, seq[i + 1] < seq[i], false);
                    // skip i-1
                }
                2 => {
                    return check_valid(seq, i + 1, i-1, decreasing, false) // skip i
                        || check_valid(seq, i, i-2, decreasing, false) // skip i-1
                        || check_valid(seq, i, i-1, seq[i] < seq[i-1], false); // skip i-2
                }
                _ => {
                    return check_valid(seq, i + 1, i-1, decreasing, false) // skip i
                        || check_valid(seq, i, i-2, decreasing, false); // skip i-1
                }
            }
        } else {
            return false;
        }
    } else {
        return check_valid(seq, i + 1, i, decreasing, can_forgive);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sample() {
        let sample = "\
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
";

        let input = parse_input(sample.to_string());

        assert_eq!(solve_part_a(&input), 2);
        assert_eq!(solve_part_b(&input), 4);
    }
}
